File: 2966_divide-array-into-arrays-with-max-difference/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> divideArray(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> subarrays;
        for (unsigned i = 0; i < nums.size(); i += 3) {
            if (i + 2 >= nums.size()) {
                vector<vector<int>> empty_subarrays;
                return empty_subarrays;
            }

            vector<int> subarray {nums[i], nums[i + 1], nums[i + 2]};
            int subarray_max = *max_element(subarray.begin(), subarray.end());
            int subarray_min = *min_element(subarray.begin(), subarray.end());

            if ((subarray_max - subarray_min) > k) {
                vector<vector<int>> empty_subarrays;
                return empty_subarrays;
            }

            subarrays.push_back(subarray);
        }

        return subarrays;
    }
};

File: 2966_divide-array-into-arrays-with-max-difference/solution.py
from typing import List

class Solution:
    def divideArray(self, nums: List[int], k: int) -> List[List[int]]:
        nums = sorted(nums)

        subarrays = [
            [
                nums[subarray_index * 3],
                nums[subarray_index * 3 + 1],
                nums[subarray_index * 3 + 2],
            ]
            for subarray_index in range(len(nums) // 3)
        ]

        for subarray in subarrays:
            if max(subarray) - min(subarray) > k:
                return []
            
        return subarrays


if __name__ == "__main__":
    Solution().divideArray([1, 1, 1, 2, 2, 2, 3, 5, 4], 1)
